Track Metadata: Fix synchronization (import/export) of file tags

uklotzde opened the pull request:
Synchronization of file tags (import/export) is severely broken due to various reasons:

- Using stale data from QFileInfo that reports outdated file modification time stamps
- Not considering *unknown* synchronization time stamps appropriately that are created by the database migration from 2.3 to 2.4. The export is skipped then, although apparently not strictly probably due to the internal caching issues caused by QFileInfo.

Those 2 different causes made it difficult to discover the misbehavior. The new unit test ensures that this will never break again.

uklotzde committed:
Add LibraryFileSyncTest

Tests the fixes from the preceding commits. Reverting each of those commits
will cause test failures.

uklotzde committed:
TrackDAO: Use correct function for reading preferences

uklotzde committed:
Fix synchronization of track metadata with file tags

uklotzde committed:
Prevent reading of stale file modification time stamps

daschuer commented on src/track/track.cpp:
```diff
@@ -1454,30 +1454,56 @@ CoverInfo Track::getCoverInfoWithLocation() const {
-        // abort
-        return ExportTrackMetadataResult::Skipped;
+    const auto sourceSyncStatus = m_record.checkSourceSyncStatus(m_fileAccess.info());
+    if (!safeToOverwriteMetadataOnExport(sourceSyncStatus)) {
```
This is hard to understand because of double inversion. 
How about 
```suggestion
    if (isLibraryMetadataOutdated(sourceSyncStatus)) {
```

Is the "Void" case correctly handled? 
When does "Void" happen? If the file has no Metadata? 
Will it trigger an assertion?

uklotzde replied:
The Void case is handled as before, the behavior didn't change.

A test for this case could be added later by someone else. I don't have any interest to do this.

uklotzde replied:
Double inversion? No, that's not the case here.

The naming of this utility function has deliberately been chosen as is, because *outdated* is just one of the possible states were exporting metadata is considered to be safe. The name of the function simply reflects the intention of a complex if-condition.

uklotzde replied:
At this point we have to make a decision if exporting metadata should be permitted or not. The function hides the implementation details, i.e. checking for outdated metadata and more.

uklotzde replied:
-> Martin Fowler - Refactoring, introductory chapters.

uklotzde replied:
I have also thought about using a verbose switch statement inside this function and am missing Rust's exhaustive match statements here. Using a switch statement inside a utility function is a valid demand. But I refuse to drop the utility function and give up the readability improvements.

daschuer commented on src/sources/metadatasource.cpp:
```diff
@@ -11,15 +11,17 @@ const Logger kLogger("MetadataSource");
 // static
-QDateTime MetadataSource::getFileSynchronizedAt(const QFileInfo& fileInfo) {
-    const QDateTime lastModifiedUtc = fileInfo.lastModified().toUTC();
+QDateTime MetadataSource::getFileSynchronizedAt(const QFile& file) {
```
This does not return a "Synchronized" time, because it can have various reason that the FileModificationTime has changed. The word "Synchronized" implies a second instance to synchronize with was there. How about:
  ```suggestion
QDateTime MetadataSource::getFileMofificationTime(const QFile& file) {
```

uklotzde replied:
I disagree. This function has been introduced do break the dependency on raw file modification time stamps. The returned value is not the actual file modification time stamp.

uklotzde replied:
Sorry, but no.

I have commented **every single line** of code in the implementation. The API comments in the header file should not reveal implementation details.

daschuer replied:
Please make an example what could be an alternative implementation. 

This issue here is that the current function name implies that Mixxx has synchronized the file. But this wit wrong. 
Conclusion: The function name is wrong.

uklotzde replied:
I disagree and refuse to make any changes.

Holzhaus replied:
In any case the method name was not changed in the PR and should not block merge. @daschuer Feel free to propose another name in a follow-up PR if you consider it necessary.

daschuer replied:
OK

daschuer commented on src/track/trackrecord.h:
```diff
@@ -115,15 +115,20 @@ class TrackRecord final {
+
+        /// The status could not be determined for whatever reason,
+        /// e.g. inaccessible file, ...
+        Undefined,
```
The difference between Undefined/Unknown/Void is not obvious in the using code. 
How about renaming these value to 
NotImported/UnknownSyncronizedTime/UnknownFileTime

uklotzde replied:
The synchronization status type should not mention files directly. It is independent of how the content of sources is stored. Files are just one kind of storage type.

As already mentioned synchronization time stamps should not be confused with raw file time stamps and we should not mix that up on this higher level of abstraction.

The names of the variants are abstract and do not need to leak implementation details from other parts of the code. Descriptions of possible use cases have deliberately been added as possible examples in the comments.

I also dislike composed terms that start with the same stem. Different kind of "unknowns" are confusing, too. It is easier to recognize distinct words.

uklotzde replied:
Please provide a mapping to the existing values.

daschuer replied:
Void = Unsynced
Syncronized = Syncronized
Outdated = ImportPending
Undefined = ExportPending

uklotzde replied:
Undefined = the source is inaccessible.

uklotzde replied:
Outdated is also somehow "unsynchronized". I don't think that this renaming improves anything.

Holzhaus replied:
I think the names in the current state of this PR are okay, and in case something is unclear there are documentation comments in the header file.

daschuer replied:
At least there is room for improvement for Undefined/Unknown/Void. In the best case the code should be self explaining without referring the header files for comment.  @uklotzde whats your proposal to fix this?

uklotzde replied:
My proposal is this PR. I did not receive any alternative proposals yet that I consider an improvement.

daschuer replied:
For my feeling my recent proposal follows your demand in https://github.com/mixxxdj/mixxx/pull/4628#discussion_r785469202 
My demand is to have more descriptive names than the ambiguous Undefined/Unknown/Void.
What would be a  alternative for you?

uklotzde replied:
The variant names are recognizable and they were helpful while writing the different test cases.

They represent a state and are not directly related to any import/export activity as you suggested.

daschuer replied:
If you really insist to keep these empty and interchangeable words Undefined/Unknown/Void and like to rely on out of sight comments, I'll give up. It is a pity that we have not found a compromise.

uklotzde replied:
Naming can be changed at any time. But I refuse to overload this PR with more unrelated changes.

daschuer reviewed (changes requested):
Thank you. I have left some comments.

uklotzde committed:
Add test for re-importing outdated metadata

uklotzde commented on src/track/track.cpp:
```diff
@@ -1454,30 +1454,56 @@ CoverInfo Track::getCoverInfoWithLocation() const {
+                    << getLocation()
+                    << "has last been been modified at"
+                    << fileSynchronizedAt
+                    << "after the last synchronization at"
```
This would be incorrect. The synchronization time stamp is updated when importing **and** exporting metadata.

daschuer replied:
I have realized later the the main source of confusion is the "fileSynchronizedAt" var name. 
This can than become: 
```suggestion
                    << "after the last library synchronization at"
```

uklotzde committed:
Remove dependency on TrackDAO

uklotzde committed:
Reword log message

daschuer commented on src/sources/metadatasource.h:
```diff
@@ -20,7 +20,10 @@ class MetadataSource {
     virtual ~MetadataSource() = default;
 
-    static QDateTime getFileSynchronizedAt(const QFileInfo& fileInfo);
+    /// Get the synchronization time of the given file.
```
This comment is wrong. the function returns the file modification time, from the HDD, this is not related to any synchronization inside Mixxx.

daschuer replied:
getFileSynchronizedAt 12:00 reads as if the file has been synchonized at 12:00. But this is not the case, an external program has touched the file at 12:00 and the file is no longer synchonized.

Holzhaus replied:
> getFileSynchronizedAt 12:00 reads as if the file has been synchonized at 12:00. But this is not the case, an external program has touched the file at 12:00 and the file is no longer synchonized.

I kind of agree that the naming may be confusing, however the method name was not changed in this PR. The PR is supposed to fix a bug, and think it's unreasonable to request refactoring larger parts of the code in a bugfix PR.

uklotzde replied:
Thank you, Jan.

daschuer replied:
OK

uklotzde committed:
Rename utility function and use an exhaustive switch/case statement

daschuer commented on src/track/track.cpp:
```diff
@@ -1454,30 +1454,70 @@ CoverInfo Track::getCoverInfoWithLocation() const {
+        // might fail subsequently. We need to continue here for avoiding
+        // misleading or wrong error messages about the synchronization status!
+        return true;
+    default:
```
Outdated is missing

uklotzde replied:
Adding a regression test for the missing enum variant revealed a conceptual problem that needs to be fixed.

uklotzde replied:
https://github.com/mixxxdj/mixxx/pull/4628/commits/ab2df29222468b71161acb28e2d74ef53cd7cda3

The comment in `Track::exportMetadata()` for the `Outdated` case explains why we have no choice other than overwriting the (maybe newer) metadata in the file and why this is a reasonable decision.

uklotzde replied:
It seems like metadata is not saved in the database if the file has disappeared (sync status = `Undefined`), regardless of the import/export synchronization settings. I had to disable the corresponding tests because this is a completely different issue.

uklotzde replied:
I have changed the behavior if export of metadata failed in https://github.com/mixxxdj/mixxx/pull/4628/commits/ffdddcb3f2e30bc5f28409fd9894822f16bb01dc. In this case the time stamp is now reset such that the sync status becomes `Unknown` when the file reappears in the future. That's the best we could do here right now.

uklotzde committed:
Fix export of outdated metadata and add tests for many use cases

uklotzde committed:
Only consider imported metadata if import succeeded

uklotzde committed:
Enable, fix, and test saving and metadata export of missing track

uklotzde committed:
LibraryFileSyncTest: Replace debug assertions with gtest macros

uklotzde committed:
LibraryFileSyncTest: Set track property directly

uklotzde committed:
Reset source sync time stamp after metadata export failed

The metadata can no longer be considered as synchronized with the source
after the export has failed and modified track metadata is saved only
in the database.

uklotzde committed:
LibraryFileSyncTest: Sleep after updating files

uklotzde commented:
I hope that https://github.com/mixxxdj/mixxx/pull/4628/commits/c72115fbdb68848b03ed08cf80077724bb0c5bb0 fixes the sporadic CI failures. One sleep might have been missing.

Holzhaus commented:
Now clang-tidy fails.

daschuer commented:
```
rror: /home/runner/work/mixxx/mixxx/src/library/trackcollectionmanager.cpp:333:42: error: this statement may fall through [-Werror=implicit-fallthrough=]
  333 |             pTrack->markForMetadataExport();
````

uklotzde committed:
Merge branch 'main' of git@github.com:mixxxdj/mixxx.git into trackrecord

uklotzde commented:
> ```
> rror: /home/runner/work/mixxx/mixxx/src/library/trackcollectionmanager.cpp:333:42: error: this statement may fall through [-Werror=implicit-fallthrough=]
>   333 |             pTrack->markForMetadataExport();
> ```

Thanks. Force pushed to repair the broken commit where the break statement got lost.

https://github.com/mixxxdj/mixxx/pull/4628/commits/bfdf55d147e165b00d6810cdc7374d84b3cf12af

uklotzde committed:
LibraryFileSyncTest: Add missing sleep after creating the temp file

uklotzde committed:
LibraryFileSyncTest: Simplify updateFileLastModified()

uklotzde committed:
LibraryFileSyncTest: Add log message for analyzing CI test failures

uklotzde commented:
The system clocks of the CI platforms don't seem to be strictly monotonic or clocks used by Qt and the file system use different sources:

```
warning [0x556a234dacd0] Track - Exporting and overwriting metadata into file "/tmp/mixxx-test-NIGwGZ/cover-test-jpg.mp3" even though the file has been modified at QDateTime(2022-01-19 00:04:12.854 UTC Qt::UTC) after the last library synchronization at QDateTime(2022-01-19 00:04:12.841 UTC Qt::UTC)
warning [0x556a234dacd0] Unexpected file modification times: after= QDateTime(2022-01-19 00:04:12.853 UTC Qt::LocalTime) <= QDateTime(2022-01-19 00:04:12.854 UTC Qt::LocalTime) = before
/home/runner/work/mixxx/mixxx/src/test/libraryfilesynctest.cpp:290: Failure
Value of: !fileLastModifiedAfter.isValid() || fileLastModifiedAfter > fileLastModifiedBefore
  Actual: false
Expected: true
```

The time stamp generated when saving the file (generated by the OS) is 1 ms behind the previous time stamp that has been generated by Qt.

I will try to increase the sleep time from 1 ms to 5 ms.

uklotzde committed:
LibraryFileSyncTest: Increase sleep time from 1 ms to 5 ms

uklotzde committed:
LibraryFileSyncTest: Rely on time stamps generated by the OS

uklotzde commented:
Back to 1 ms with time stamps generated by the OS. Let's see if this works out.

uklotzde committed:
LibraryFileSyncTest: Split into separate test classes

uklotzde committed:
LibraryFileSyncTest: Add missing verifications

uklotzde committed:
LibraryFileSyncTest: Move shared code into base class

uklotzde committed:
LibraryFileSyncTest: Add tests for files without metadata

daschuer commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,712 @@
+        const auto newLastModified = fileLastModified();
+        ASSERT_TRUE(newLastModified.isValid());
+        ASSERT_GT(newLastModified, oldLastModified);
+        sleepAfterFileLastModifiedUpdated();
```
Calling sleep in a function called updateFileLastModified() is supprising. 
I think this should be moved to the outer scope to understand the issue with > and =. 
Or can we dispose this?

daschuer replied:
This also applies to all other invocations of sleepAfterFileLastModifiedUpdated();
If this is essential to test is shall be called in the test to understand the logic in place.

uklotzde replied:
The test cases should focus on the business logic and not be bothered with technical workarounds. That's one purpose of those utility classes.

uklotzde replied:
https://github.com/mixxxdj/mixxx/pull/4628#discussion_r788003336

daschuer replied:
I disagree here. The user should know in the test what he has to expect from the timing. This is essential, because the test may fail because of timing issues.

uklotzde commented:
As a last step before merging I would like to rename this test from `LibraryFileSyncTest` to `TrackMetadataSyncTest`, because that's the feature that actually gets tested here.

uklotzde committed:
LibraryFileSyncTest: Improve type safety and naming

daschuer commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,712 @@
+        pTrack.reset();
+        ASSERT_EQ(nullptr, GlobalTrackCacheLocker().lookupTrackById(trackId));
+        // The file might have been modified if track metadata has been exported.
+        sleepAfterFileLastModifiedUpdated();
```
Can this be removed? It looks like it is wrong here.

uklotzde replied:
I don't think so. We have to insert a sleep whenever file system time stamps are modified to ensure... Please read my code comments. I am not going to repeat them here.

daschuer replied:
I don't understand the comment here. Can you improve it? 

It is used in two cases, an in both cases it does not shift following times. it just delay the lookup of the just written file time. 

If this has an effect something else must go on. Maybe is a aching issue or such.

uklotzde replied:
The sleep only affects any following write operation. Reading the modified time stamp does not need to be delayed, but the sleep delays everything.

The comment says exactly what happens, i.e. a write operation to the file system might happen behind the scenes depending on the circumstances.

uklotzde replied:
Caching issues have already been fixed and file times are always read directly in the test.

daschuer commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,712 @@
+                testFile.absoluteFilePath(),
+                m_fileInfo.location());
+        ASSERT_TRUE(m_fileInfo.exists());
+        sleepAfterFileLastModifiedUpdated();
```
Please move that to the constructor or rename the function to createFileAndSleep() or createAndAgeFile() or such.

uklotzde replied:
Already tried that. Not possible due to the gtest macros.

daschuer replied:
It looks like you answer belongs to a different thread.  

Here I have asked to move the sleep into the calling function or rename the function.

uklotzde replied:
No. The time machinery behind the scenes doesn't belong to the actual testing code. It is just there to ensure that all generated time stamps that are generated in quick succession during the tests are strictly monotonic.

This is the last time I am trying to explain it.

daschuer commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,712 @@
+        const auto pTrack = getOrAddTrackByLocation(trackRef.getLocation());
+        ASSERT_TRUE(trackCollectionManager()->getTrackByRef(trackRef));
+        m_trackId = pTrack->getId();
+        sleepAfterFileLastModifiedUpdated();
```
Why we need this delay? We already have such a delay when creating the file. So the track is already kind of aged.  
I think that should be moved into the tests to make it obvious why we need it.

uklotzde replied:
This one seems indeed redundant and a leftover from a previous rewrite. Will try.

uklotzde commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,712 @@
+    void establishSourceSyncStatus(TrackPointer* pLoadedTrack) const {
+        // Touch the file's modification time stamp to simulate an external
+        // modification by a 3rd party app after the track has been loaded.
+        updateFileLastModified();
```
I don't think so. Let's leave it as is and you may try later if you have the time to do so.

uklotzde replied:
We cannot reorder the code because the database includes the check of the desired status that would fail if we modify the time stamp later.

uklotzde replied:
updateFileLastModifiedAndSleep(): This naming would be wrong, because it reveals an implementation detail that should be hidden from the caller.

Once again: The invocation of sleep does not belong to the test routine! It is only required for technical reasons because we have no control of the side effects cause by the file system access.

daschuer replied:
A function should be named like what it does and not hide important information. But I prefer anyway to move all the sleeps one level above. That would make the code much easier to read and the test more consistent.

uklotzde commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,712 @@
+        verifySourceSyncStatusOfTrack(*pTrack, expectedSourceSyncStatus);
+
+        if (pLoadedTrack) {
+            *pLoadedTrack = std::move(pTrack);
```
Not possible when using the gtest macros.

daschuer replied:
Please use EXPECT_ than it works ASSERT_ does not work because the return only form the function.

uklotzde replied:
https://github.com/mixxxdj/mixxx/pull/4628/commits/a4f550ba80c3e2af1058bf5f4bcedf8cd5df1186

uklotzde commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,712 @@
+    }
+
+  private:
+    void establishSourceSyncStatus(TrackPointer* pLoadedTrack) const {
```
No. This is already part of the class name. The member functions are named identical in all derived classes.

daschuer replied:
That's no reason for giving private member function bad names.

uklotzde replied:
I disagree and will not change the naming of the member functions. This would affect all other member functions too, would produce a lot of noise and would also change the consistent implementation code of the individual tests. Currently these tests can be created simply by copy&paste because of the identical naming!

daschuer replied:
copy & past is evil :-)  

Seriously, that is no reason for bad function names!
You may consider to manual inline them as an alternative. They are called from one place only anyway.

uklotzde replied:
Daniel, seriously. Then please rewrite everything in your style and don't bother me anymore.

I am well aware that the code is still far from ideal. I have already spent a lot of time to come up with these generic testing functions. On the other hand you are asking me to remove the generic functions and replace everything with even more duplicated, boilerplate code [here](https://github.com/mixxxdj/mixxx/pull/4628#issuecomment-1016829290).

Be-ing replied:
@uklotzde has already invested a lot of work into fixing these complex problems that have real consequences for users. Please do not bother him about implementation details that have literally no impact on people running the code.

uklotzde replied:
This functions does not longer exist. It has been replaced by the polymorphic `TrackPointer prepareTestTrack()`.

https://github.com/mixxxdj/mixxx/pull/4628/commits/a4f550ba80c3e2af1058bf5f4bcedf8cd5df1186

uklotzde committed:
LibraryFileSyncTest: Remove stray sleepAfterFileLastModifiedUpdated()

daschuer commented:
This test fails locally. Running with ctest. Unfortunatly with no useful output. 
```
417 - LibraryFileWithoutMetadataSyncStatusOutdatedTest.checkTrackRecordSourceSyncStatus (Failed)
```

daschuer commented:
```
/home/sperry/workspace/2.3/src/test/libraryfilesynctest.cpp:151: Failure
Expected equality of these values:
  expectedSourceSyncStatus
    Which is: 4-byte object <03-00 00-00>
  trackRecord.checkSourceSyncStatus(track.getFileInfo())
    Which is: 4-byte object <02-00 00-00>
/home/sperry/workspace/2.3/src/test/libraryfilesynctest.cpp:167: Failure
Expected: (fileLastModified) > (sourceSynchronizedAt), actual: 8-byte object <4F-1F E7-DD 73-7E 01-00> vs 8-byte object <1F-9F F8-A6 73-7E 01-00>
/home/sperry/workspace/2.3/src/test/libraryfilesynctest.cpp:374: Failure
Expected equality of these values:
  expectedSourceSyncStatus
    Which is: 4-byte object <03-00 00-00>
  trackRecordOfTestTrack.checkSourceSyncStatus(pTrack->getFileInfo())
    Which is: 4-byte object <02-00 00-00>
info [0x55802df8dfd0] TrackCollection - Disconnecting database
info [0x55802df8dfd0] GlobalTrackCache - Destroying instance
[  FAILED  ] LibraryFileWithoutMetadataSyncStatusOutdatedTest.checkTrackRecordSourceSyncStatus (20 ms)
[----------] 1 test from LibraryFileWithoutMetadataSyncStatusOutdatedTest (20 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (20 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] LibraryFileWithoutMetadataSyncStatusOutdatedTest.checkTrackRecordSourceSyncStatus

 1 FAILED TEST


0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.11 sec

The following tests FAILED:
	417 - LibraryFileWithoutMetadataSyncStatusOutdatedTest.checkTrackRecordSourceSyncStatus (Failed)
Errors while running CTest
```

daschuer commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,711 @@
+    void updateFileLastModified() const {
+        auto file = m_fileInfo.toQFile();
+        const auto oldLastModified = fileLastModified();
+        ASSERT_TRUE(oldLastModified.isValid());
```
ASSERT_ macros are not (well) designed to use in subroutines. They are designed to immediately stop the test when used in the test body. Unfortunately they do not abort the whole run like normal asserts, the just return from the function, which is not desired here. You should replace them with EXPECT_ macros, for a deterministic behavior of the test assumptions in the following code. This way you also get rid of the requirement to use void functions in tests.

uklotzde replied:
The gtest ASSERT macros are used to very preconditions. These preconditions are "established" in subroutines.

daschuer replied:
Sorry, that is simply not true. 

Please use EXPECT, introduce a proper return value instead of the pointer parameter and both topics are solved.

uklotzde replied:
If preconditions are violated tests should be aborted immediately. That's what assertions are for. Please consult the official docs and tell me what I am misunderstanding about it:

http://google.github.io/googletest/reference/assertions.html

Holzhaus replied:
I think Daniel means that the ASSERT macros only abort the current function, not the whole test (which would be the right thing to do):

> Upon failure, `EXPECT_` macros generate nonfatal failures and allow the current function to continue running, while `ASSERT_` macros generate fatal failures and abort the current function.

But I'm actually now sure how that is supposed to work. I guess the documentation is misleading. Otherwise I have no idea what would happen if the function had a return value?

Holzhaus replied:
If that is indeed the case, then `EXPECT` would also be wrong and we should make the hlper function return a bool instead, and assert that the output value is true in the actual test case.

uklotzde replied:
It was returning a boolean before but that has proven to be tedious to use. I'll check again.

uklotzde replied:
Returning a boolean would require a return statement after every EXPECT macro. I was  using VERIFY_AND_DEBUG_ASSERT for this purpose but that only works in debug mode and does not indicate the reason for the failure when compiling with DEBUG_ASSERTIONS_FATAL=OFF. It is not suitable for testing.

uklotzde replied:
I am not going to write a new testing framework to overcome the deficiencies of GoogleTest.

uklotzde replied:
https://github.com/mixxxdj/mixxx/pull/4628/commits/a4f550ba80c3e2af1058bf5f4bcedf8cd5df1186

daschuer commented:
There seems to be another issue when running ./mixxx_test directly: 
```
./mixxx-test --gtest_filter=LibraryFile*
QML debugging is enabled. Only use this in a safe environment.
Note: Google Test filter = LibraryFile*
[==========] Running 26 tests from 8 test suites.
[----------] Global test environment set-up.
[----------] 3 tests from LibraryFileWithMetadataSyncStatusSynchronizedTest
[ RUN      ] LibraryFileWithMetadataSyncStatusSynchronizedTest.checkTrackRecordSourceSyncStatus
[       OK ] LibraryFileWithMetadataSyncStatusSynchronizedTest.checkTrackRecordSourceSyncStatus (45 ms)
[ RUN      ] LibraryFileWithMetadataSyncStatusSynchronizedTest.saveTrackMetadataWithSyncEnabled
critical [0x55fda9d03fd0] DEBUG ASSERT: "!pProviderByDisplayName" in function int mixxx::SoundSourceProviderRegistry::registerProvider(const SoundSourceProviderPointer&) at /home/sperry/workspace/2.3/src/sources/soundsourceproviderregistry.cpp:53
```

daschuer commented:
sometimes 
	419 - LibraryFileWithoutMetadataSyncStatusOutdatedTest.saveTrackMetadataWithSyncDisabled (Failed)

uklotzde commented:
> There seems to be another issue when running ./mixxx_test directly:
> 
> ```
> ./mixxx-test --gtest_filter=LibraryFile*
> QML debugging is enabled. Only use this in a safe environment.
> Note: Google Test filter = LibraryFile*
> [==========] Running 26 tests from 8 test suites.
> [----------] Global test environment set-up.
> [----------] 3 tests from LibraryFileWithMetadataSyncStatusSynchronizedTest
> [ RUN      ] LibraryFileWithMetadataSyncStatusSynchronizedTest.checkTrackRecordSourceSyncStatus
> [       OK ] LibraryFileWithMetadataSyncStatusSynchronizedTest.checkTrackRecordSourceSyncStatus (45 ms)
> [ RUN      ] LibraryFileWithMetadataSyncStatusSynchronizedTest.saveTrackMetadataWithSyncEnabled
> critical [0x55fda9d03fd0] DEBUG ASSERT: "!pProviderByDisplayName" in function int mixxx::SoundSourceProviderRegistry::registerProvider(const SoundSourceProviderPointer&) at /home/sperry/workspace/2.3/src/sources/soundsourceproviderregistry.cpp:53
> ```

Unrelated. These failure happen if you don' use ctest.

```
ctest \
--output-on-failure \
--stop-on-failure \
-j 4 \
--tests-regex "LibraryFile.*Test" \
```

daschuer commented:
Fore my feeling the new test class overshoots our goal by far. We have no interest in tests, testing file timing in a separate testing calls. The test should test the productive code. In the productive code we have a > operator for testing for outdated files. We have no case where we write a file and read it back immediately. So the implemented ms resolution is not required. 

The only reason I can consider in real live where this is relevant is a concurrent write to the same file form Mixxx and another Tool. Here is the result undefined anyway.  

It should be sufficient to test this with a reasonable old file. So you may consider to reintroduce the changing the file time directly and set it to a age of one hour or such. This way we can get rid of all the tests failures related to file time resolution and caching. 

I also prefer to go back to a procedural test body that is closer to the implementations of other tests. This is way easier to read instead of your deeply nested test classes where the control flow is hard to track, especially because an implicit run time measurement is involved.

uklotzde commented:
> ```
> /home/sperry/workspace/2.3/src/test/libraryfilesynctest.cpp:151: Failure
> Expected equality of these values:
>   expectedSourceSyncStatus
>     Which is: 4-byte object <03-00 00-00>
>   trackRecord.checkSourceSyncStatus(track.getFileInfo())
>     Which is: 4-byte object <02-00 00-00>
> /home/sperry/workspace/2.3/src/test/libraryfilesynctest.cpp:167: Failure
> Expected: (fileLastModified) > (sourceSynchronizedAt), actual: 8-byte object <4F-1F E7-DD 73-7E 01-00> vs 8-byte object <1F-9F F8-A6 73-7E 01-00>
> /home/sperry/workspace/2.3/src/test/libraryfilesynctest.cpp:374: Failure
> Expected equality of these values:
>   expectedSourceSyncStatus
>     Which is: 4-byte object <03-00 00-00>
>   trackRecordOfTestTrack.checkSourceSyncStatus(pTrack->getFileInfo())
>     Which is: 4-byte object <02-00 00-00>
> info [0x55802df8dfd0] TrackCollection - Disconnecting database
> info [0x55802df8dfd0] GlobalTrackCache - Destroying instance
> [  FAILED  ] LibraryFileWithoutMetadataSyncStatusOutdatedTest.checkTrackRecordSourceSyncStatus (20 ms)
> [----------] 1 test from LibraryFileWithoutMetadataSyncStatusOutdatedTest (20 ms total)
> 
> [----------] Global test environment tear-down
> [==========] 1 test from 1 test suite ran. (20 ms total)
> [  PASSED  ] 0 tests.
> [  FAILED  ] 1 test, listed below:
> [  FAILED  ] LibraryFileWithoutMetadataSyncStatusOutdatedTest.checkTrackRecordSourceSyncStatus
> 
>  1 FAILED TEST
> 
> 
> 0% tests passed, 1 tests failed out of 1
> 
> Total Test time (real) =   0.11 sec
> 
> The following tests FAILED:
> 	417 - LibraryFileWithoutMetadataSyncStatusOutdatedTest.checkTrackRecordSourceSyncStatus (Failed)
> Errors while running CTest
> ```

I remember to have seen this failure on CI (Ubuntu) after setting the sleep time to 0 ms. Did you try to increase the minimum sleep time of 1 ms?

I cannot reproduce timing issues locally. All tests succeed on every run.

daschuer commented:
I run the original PR merged to master.

uklotzde commented:
> Fore my feeling the new test class overshoots our goal by far. We have no interest in tests, testing file timing in a separate testing calls. The test should test the productive code. In the productive code we have a > operator for testing for outdated files. We have no case where we write a file and read it back immediately. So the implemented ms resolution is not required.
> 
> The only reason I can consider in real live where this is relevant is a concurrent write to the same file form Mixxx and another Tool. Here is the result undefined anyway.
> 
> It should be sufficient to test this with a reasonable old file. So you may consider to reintroduce the changing the file time directly and set it to a age of one hour or such. This way we can get rid of all the tests failures related to file time resolution and caching.
> 
> I also prefer to go back to a procedural test body that is closer to the implementations of other tests. This is way easier to read instead of your deeply nested test classes where the control flow is hard to track, especially because an implicit run time measurement is involved.

These extensive tests with all their combinatorics have revealed many of issues.

uklotzde commented:
We should restart the build jobs repeatedly to see if there are still timing issues on Ubuntu with the minimum delay of 1 ms between file modifications.

All timing code in Mixxx and the tests should only compare time stamps that have been generated by the file system as the single source of truth. I am not aware that we missed any occurrence that uses time stamps from a second, inconsistent source.

Be-ing commented:
Manually restarting GitHub Actions jobs will cancel the old run. You can hack around this by repeatedly doing `git commit --amend` and force pushing to create jobs testing identical code with different commit hashes.

uklotzde commented:
> Manually restarting GitHub Actions jobs will cancel the old run. You can hack around this by repeatedly doing `git commit --amend` and force pushing to create jobs testing identical code with different commit hashes.


> Manually restarting GitHub Actions jobs will cancel the old run. You can hack around this by repeatedly doing `git commit --amend` and force pushing to create jobs testing identical code with different commit hashes.

Restarting works as expected and the tests have been rerun as intended. We want them to fail, not to succeed.

uklotzde committed:
Replace ASSERT with EXPECT and restore function signatures

uklotzde committed:
Loop until a file is detected as modified

uklotzde committed:
Insert a sleep before and not after file modifications

This saves 1 sleep operation per test run.

uklotzde commented:
https://github.com/mixxxdj/mixxx/pull/4628/commits/ed153113fcf01d5ce1c97b424a244402308e4f74 will hopefully fix the Ubuntu issues.

https://github.com/mixxxdj/mixxx/pull/4628/commits/78ecbac427465a77686285338851406f1387c8be saves 1 sleep operation per test case.

uklotzde committed:
Merge branch 'main' of git@github.com:mixxxdj/mixxx.git into trackrecord

daschuer commented:
It is still failing locally: 
```
info [0x55bd0389efd0] TrackCollectionManager - External collections are disabled in test mode
info [0x55bd0389efd0] TrackCollectionManager - Library scanner is disabled in test mode
info [0x55bd0389efd0] SeratoMarkers2 - Reading bpmlock state from 'Serato Markers2' tag data...
info [0x55bd0389efd0] SeratoMarkers2 - Reading track color from 'Serato Markers2' tag data...
warning [0x55bd0389efd0] Track - Exporting and overwriting metadata into file "/tmp/mixxx-test-NjCcZd/cover-test-jpg.mp3" even though the file has been modified at QDateTime(2022-01-20 11:29:18.314 UTC Qt::UTC) after the last library synchronization at QDateTime(2022-01-20 11:29:18.306 UTC Qt::UTC)
/home/sperry/workspace/2.3/src/test/libraryfilesynctest.cpp:253: Failure
Value of: !fileLastModifiedAfter.isValid() || fileLastModifiedAfter > fileLastModifiedBefore
  Actual: false
Expected: true
info [0x55bd0389efd0] TrackCollection - Disconnecting database
info [0x55bd0389efd0] GlobalTrackCache - Destroying instance
[  FAILED  ] LibraryFileWithMetadataSyncStatusOutdatedTest.saveTrackMetadataWithSyncEnabled (24 ms)
[----------] 1 test from LibraryFileWithMetadataSyncStatusOutdatedTest (24 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (24 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] LibraryFileWithMetadataSyncStatusOutdatedTest.saveTrackMetadataWithSyncEnabled

 1 FAILED TEST
```

uklotzde commented:
> It is still failing locally in rare cases:
> 
> ```
> info [0x55bd0389efd0] TrackCollectionManager - External collections are disabled in test mode
> info [0x55bd0389efd0] TrackCollectionManager - Library scanner is disabled in test mode
> info [0x55bd0389efd0] SeratoMarkers2 - Reading bpmlock state from 'Serato Markers2' tag data...
> info [0x55bd0389efd0] SeratoMarkers2 - Reading track color from 'Serato Markers2' tag data...
> warning [0x55bd0389efd0] Track - Exporting and overwriting metadata into file "/tmp/mixxx-test-NjCcZd/cover-test-jpg.mp3" even though the file has been modified at QDateTime(2022-01-20 11:29:18.314 UTC Qt::UTC) after the last library synchronization at QDateTime(2022-01-20 11:29:18.306 UTC Qt::UTC)
> /home/sperry/workspace/2.3/src/test/libraryfilesynctest.cpp:253: Failure
> Value of: !fileLastModifiedAfter.isValid() || fileLastModifiedAfter > fileLastModifiedBefore
>   Actual: false
> Expected: true
> info [0x55bd0389efd0] TrackCollection - Disconnecting database
> info [0x55bd0389efd0] GlobalTrackCache - Destroying instance
> [  FAILED  ] LibraryFileWithMetadataSyncStatusOutdatedTest.saveTrackMetadataWithSyncEnabled (24 ms)
> [----------] 1 test from LibraryFileWithMetadataSyncStatusOutdatedTest (24 ms total)
> 
> [----------] Global test environment tear-down
> [==========] 1 test from 1 test suite ran. (24 ms total)
> [  PASSED  ] 0 tests.
> [  FAILED  ] 1 test, listed below:
> [  FAILED  ] LibraryFileWithMetadataSyncStatusOutdatedTest.saveTrackMetadataWithSyncEnabled
> 
>  1 FAILED TEST
> ```

Unfortunately we cannot loop our track saving code until a different time stamp for the file is generated when exporting metadata. Then we need to increase the sleep time until the tests succeed reliably.

uklotzde committed:
Testing: Increase sleep time for file modification time stamps

daschuer commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,701 @@
+        // The file might be modified if track metadata is exported.
+        // Ensures that the resulting time stamp will be strictly
+        // greater than the current one.
+        sleepBeforeUpdatingFileLastModifiedTime();
```
The sleepBeforeUpdatingFileLastModifiedTime here might be insufficient? 

The sleep belongs to the context of the test where a > detects a file write. The best would be to age the original file at the beginning of the outer function.  Alternatively you may to rename the function to a name that reveals this nature. 
saveModifiedTrackAndAdvanceFileTime();

uklotzde replied:
Saving a track is implicitly supposed to touch the file depending on the sync settings.

Consequently following your proposal I would have to rename it to saveModifiedTrackAndMaybeExportFileTagsAndAdvanceFileTime() or stuff even more implementation details into the function name. This is not helpful.

daschuer commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,701 @@
+        ASSERT_EQ(1, pTrack.use_count());
+        const auto trackId = pTrack->getId();
+        ASSERT_TRUE(trackId.isValid());
+        ASSERT_TRUE(pTrack->isDirty());
```
Is this the case? This does not match to the comment below.

uklotzde replied:
What do you mean?

daschuer commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,701 @@
+    WithoutMetadata,
+};
+
+void sleepBeforeUpdatingFileLastModifiedTime() {
```
Functions should be named for what they do, not depending on the outer context. 
So this could be advanceFileTimeSleep() or such.

uklotzde replied:
The naming here should reveal where the function is ought to be used and has been chosen deliberately.

This is **NOT** a regular function as in regular code. It has been created for documentation purposes.

daschuer commented:
The latest commit works here. However I would prefer a more robust solution by just setting the file time to an reasonable old value on creation.

uklotzde commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,696 @@
+    }
+
+    void removeFile() const {
+        ASSERT_TRUE(m_fileInfo.toQFile().remove());
```
Why? ASSERT is semantically correct here and we can use it because of the `void` return.

daschuer replied:
An asset violation normally applied a fatal error. But the Google test assert macro only works in the function body for that purpose. 

Here the function ends normally. So EXPECT is less confusing.

uklotzde replied:
https://github.com/mixxxdj/mixxx/pull/4628/commits/186d9c1df31bd66f4ed6879518aee64021dc64f0

uklotzde commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,696 @@
+        // time stamps.
+        auto newLastModified = oldLastModified;
+        do {
+            sleepBeforeUpdatingFileLastModifiedTime();
```
No need to complicate the code that is almost never executed. It doesn't hurt to sleep repeatedly instead of spinning in a hot loop.

uklotzde replied:
ASSERT is semantically correct, see below.

daschuer replied:
Google asserts do don't work for the implied semantic. For that purpose you need to use normal asserts. But we want the test output. So please use ECPECT to not give wrong hints die to a not working semantic.

uklotzde replied:
https://github.com/mixxxdj/mixxx/pull/4628/commits/186d9c1df31bd66f4ed6879518aee64021dc64f0

uklotzde committed:
Rename function in test code

uklotzde committed:
Prevent infinite loop in test code

uklotzde commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,696 @@
+        return m_fileInfo.toQFile().fileTime(QFileDevice::FileModificationTime);
+    }
+
+    void updateFileLastModified() const {
```
https://github.com/mixxxdj/mixxx/pull/4628/commits/23e623f2d5795643b23e1faa35cf2dc4d5dc930f

uklotzde commented on src/test/libraryfilesynctest.cpp:
```diff
@@ -0,0 +1,696 @@
+        // sequences of time stamps when mixed with the file system
+        // time stamps.
+        auto newLastModified = oldLastModified;
+        do {
```
https://github.com/mixxxdj/mixxx/pull/4628/commits/082cc7edb6e36153be17cbca83dabeb0133558df

uklotzde committed:
LibraryFileSyncTest: Replace all ASSERT_ with EXPECT_ macros

uklotzde commented:
> The latest commit works here. However I would prefer a more robust solution by just setting the file time to an reasonable old value on creation.

That would be inconsistent with the implicit update of the modification time stamp when saving the file. We don't know anything about the relation of the different clocks and should rely on a single source of truth.

daschuer commented:
Original we had in one case a bussy loop that finds exactly the minimum difference that is considered as such. But you had no solution for setting the file to a newer value. 

Now you have increased the sleep time to 2 ms, a wild guess that works on my device which makes the busy loop obsolete, because the 2 is used there as well. 

This mean you original demand to test with them I um difference does no longer work anyway.

I think a much better solution is to creat the file, read the file time and sibstract 1 and write it back.

The same can be done when advancing the file time read it add one and write it back. This way it should be rock solid.

uklotzde committed:
Adjust file times instead of sleeping for metadata sync tests

uklotzde commented:
> Original we had in one case a bussy loop that finds exactly the minimum difference that is considered as such. But you had no solution for setting the file to a newer value.
> 
> Now you have increased the sleep time to 2 ms, a wild guess that works on my device which makes the busy loop obsolete, because the 2 is used there as well.
> 
> This mean you original demand to test with them I um difference does no longer work anyway.
> 
> I think a much better solution is to creat the file, read the file time and sibstract 1 and write it back.
> 
> The same can be done when advancing the file time read it add one and write it back. This way it should be rock solid.

Thanks. That was actually a good idea. The current test design has proven to be solid as this change was easy to integrate. Though somehow also tricky, I added comments.

uklotzde committed:
Rename enum class

uklotzde committed:
Add comment about test execution

uklotzde committed:
Rename unit test

uklotzde committed:
Merge branch 'main' of git@github.com:mixxxdj/mixxx.git into trackrecord

daschuer commented:
Now it is working reliable. Thank you. I am sorry for you burned fingers.
